src/engine.ml

```diff
@@ -525,10 +533,18 @@ let tls_ekm tls config =
   | None, Some flags -> if List.mem `Tls_ekm flags then Some tls else None
   | None, None -> None
 
-let kex_server config session (my_key_material : my_key_material) tls data =
+let kex_server config auth_user_pass session (my_key_material : my_key_material)
+    tls data =
   let open Result.Syntax in
-  (* TODO verify username + password, respect incoming data *)
   let* their_tls_data = Packet.decode_tls_data ~with_premaster:true data in
+  let* () =
+    match (auth_user_pass, their_tls_data.user_pass) with
+    | None, _ -> Ok ()
+    | Some _, None -> Error (`Msg "no username and password provided")
+    | Some auth, Some (user, pass) ->
+        if auth ~user ~pass then Ok ()
+        else Error (`Msg "couldn't verify username and password")
```

Comment:
Hmm. Should we would send `AUTH_FAILED` here? I'm not so sure about using `Error _` here.

Reply:
The openvpn-rfc is confusing wrt. `AUTH_FAILED`, and the section on user/password authentication is [TBD](https://github.com/OpenVPN/openvpn-rfc/blob/master/openvpn-wire-protocol.xml#L1967-L1970)...

Reply:
how should that be sent? I think we should send it to the client, but I was not aware how the encoding / on which channel this should be sent.

Reply:
thanks for your investigations and fixes, this is done in later commits ([d5e1be4](https://github.com/robur-coop/miragevpn/pull/249/commits/d5e1be42e762c85a1dd91b2ede1df7d37329facb) [37e4832](https://github.com/robur-coop/miragevpn/pull/249/commits/37e483239eff6e394b04e78b971509c58ad5647a))